### PR TITLE
Add tests and docs to product function

### PR DESCRIPTION
**Pull request by pineapplemachine**
Also did some misc minor polish tasks in the process, including:

Adding a `throws` attribute to some function docs objects

Adding `isNaN` and `assertNaN` functions

### CODE REVIEW

File: src/core/types.js [hunk 1]
```diff
@@ -14,6 +14,12 @@ export const isInteger = (value) => {
     return Number.isInteger(value);
 };
 
+export const isNaN = (value) => {
+    // EMCA spec says value !== value is true if and only if value is NaN.
+    // https://tc39.github.io/ecma262/#sec-isnan-number
+    return value !== value;
+};
```
Comment (pineapplemachine) [inline src/core/types.js:21]: Different from builtin `isNaN` in that this returns true _only_ for `hi.isNaN(NaN)` and returns false for other non-numeric inputs

File: src/functions/any.js [hunk 2]
```diff
@@ -21,6 +21,10 @@ export const any = wrap({
             elements were truthy. The function returns @false otherwise, or
             if the sequence was empty.
         `),
+        throws: (`
+            The function throws a @NotBoundedError when the input sequence was
+            not known to be bounded.
+        `),
```
Comment (pineapplemachine) [inline src/functions/any.js:27]: All functions that may throw an error in their `implementation` function should describe the errors and the circumstances under which they are thrown in `throws` strings like this

File: src/functions/product.js [hunk 3]
```diff
@@ -1,19 +1,88 @@
...
+            "basicUsage",
+        ],
+    },
     attachSequence: true,
     async: true,
     arguments: {
         one: wrap.expecting.iterable
     },
     implementation: (source) => {
+        NotBoundedError.enforce(source, {
```
Comment (pineapplemachine) [inline src/functions/product.js:38]: bugfix

File: src/functions/product.js [hunk 4]
```diff
@@ -1,19 +1,88 @@
...
+        "singleLengthInput": hi => {
+            hi.assert(hi.product([0]) === 0);
+            hi.assert(hi.product([1]) === 1);
+            hi.assert(hi.product([2]) === 2);
+            hi.assert(hi.product([50]) === 50);
+        },
+        "numericStringInput": hi => {
+            // Weird case but it's more work to invalidate than to just accept it
+            hi.assert(hi.product([2, "3"]) === 6);
+            hi.assert(hi.product(["5", "6"]) === 30);
```
Comment (pineapplemachine) [inline src/functions/product.js:67]: This being a valid pattern makes me squirm so hard but I don't think it's wise to add extra checks to the loop to invalidate this

File: src/functions/product.js [hunk 5]
```diff
@@ -1,19 +1,88 @@
...
+        },
+        "nanInput": hi => {
+            hi.assertNaN(hi.product(["not a number"]));
+            hi.assertNaN(hi.product([1, 2, 3, "?"]));
+        },
+        "notKnownBoundedInput": hi => {
+            hi.assertFail(
+                (error) => (error.type === "NotBoundedError"),
+                () => hi.recur(i => i + 1).seed(0).until(i => i === 10).product()
```
Comment (pineapplemachine) [inline src/functions/product.js:76]: Ahhh there needs to be an easier way for testing to get some arbitrary sequence that isn't known-bounded or known-unbounded

For an unbounded sequence you can just `counter` but this is kind of tedious

Comment (jpsullivan) [reply src/functions/product.js:76]: The [ava](https://github.com/avajs/ava) people do the following to test for exception throwing:
```javascript
test('throws', t => {
	const error = t.throws(() => {
		fn();
	}, TypeError);

	t.is(error.message, '🦄');
});
```

Dunno if that helps at all, but maybe it's something we can work towards.

**Review by jpsullivan — APPROVED** [review]

**Pull request merged**